onchain/rollups/contracts/consensus/quorum/Quorum.sol

```diff
@@ -0,0 +1,152 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+pragma solidity ^0.8.8;
+
+import {BitMaps} from "@openzeppelin/contracts/utils/structs/BitMaps.sol";
+
+import {AbstractConsensus} from "../AbstractConsensus.sol";
+import {InputRange} from "../../common/InputRange.sol";
+
+/// @notice A consensus model controlled by a small, immutable set of `n` validators.
+/// @notice You can know the value of `n` by calling the `numOfValidators` function.
+/// @notice Upon construction, each validator is assigned a unique number from 1 and `n`.
+/// These numbers are used internally instead of addresses for optimization reasons.
+/// @notice You can list the validators in the quorum by calling the `validatorById`
+/// function for each ID from 1 to `n`.
+contract Quorum is AbstractConsensus {
+    using BitMaps for BitMaps.BitMap;
+
+    /// @notice The total number of validators.
+    /// @notice See the `numOfValidators` function.
+    uint256 private immutable _numOfValidators;
+
+    /// @notice Validator IDs indexed by address.
+    /// @notice See the `validatorId` function.
+    /// @dev Non-validators are assigned to ID zero.
+    /// @dev Validators have IDs greater than zero.
+    mapping(address => uint256) private _validatorId;
+
+    /// @notice Validator addresses indexed by ID.
+    /// @notice See the `validatorById` function.
+    /// @dev Invalid IDs are assigned to address zero.
+    mapping(uint256 => address) private _validatorById;
+
+    /// @notice Votes in favor of a particular claim.
+    /// @param inFavorCount The number of validators in favor of the claim
+    /// @param inFavorById The set of validators in favor of the claim
+    /// @dev `inFavorById` is a bitmap indexed by validator IDs.
+    struct Votes {
+        uint256 inFavorCount;
+        BitMaps.BitMap inFavorById;
+    }
+
+    /// @notice Votes indexed by claim.
+    /// @dev See the `numOfValidatorsInFavorOf` and `isValidatorInFavorOf` functions.
+    mapping(address => mapping(uint256 => mapping(uint256 => mapping(bytes32 => Votes))))
+        private _votes;
+
+    /// @param validators The array of validator addresses
+    /// @dev Duplicates in the `validators` array are ignored.
```

Comment:
Is it worth to pay the gas for this logic to ignore repeated validators in the array? Maybe this can be handled by the script/frontend/individual that calls the QuorumFactory.

The consequences of repeated validators is a broken Quorum, but it shouldnt affect anyone unless a DApp dont know its broken.

Reply:
Given that the quorum will be generally small, and that this code will only be executed once by whoever deploys the `Quorum` contract, I think it is worth.

Reply:
After some benchmarking, it seems that we are only wasting 0.1% more gas, so I think it is still worth it.